### PR TITLE
Test front

### PR DESCRIPTION
**Pull request by yliu0715**
- Add Unit tests for React
- Refactor some of the fetch calls to use async/await

### CODE REVIEW

File: src/web/client/src/components/Navbar/Navbar.test.tsx [hunk 1]
```diff
@@ -0,0 +1,55 @@
...
+    const onAuthChangeMock: jest.Mock = jest.fn();
+    const { container } = render(
+      <Router>
+        <Navbar isLoggedIn={false} onAuthChange={onAuthChangeMock} />
+      </Router>,
+    );
+    expect(container.firstChild).toMatchSnapshot();
+  });
+
+  // will get network request failed error if this test is put right above
```
Comment (melvinheks) [inline src/web/client/src/components/Navbar/Navbar.test.tsx:18]: lmao

File: src/web/client/src/components/LiveStream/LiveStream.test.tsx [hunk 2]
```diff
@@ -0,0 +1,45 @@
+import React from "react";
+import { shallow } from "enzyme";
+import { render } from "@testing-library/react";
+import LiveStream from "./LiveStream";
+
+describe("LiveStream component", () => {
+  let mockCollection: object;
+  let mockJsonPromise: any;
+  let mockFetchPromise: any;
```
Comment (chc5) [inline src/web/client/src/components/LiveStream/LiveStream.test.tsx:9]: Shouldn't the type be Promise<object> and Promise?

File: src/web/client/src/components/Login/Login.tsx [hunk 3]
```diff
@@ -41,30 +40,31 @@ export default class Login extends Component<Props, State> {
...
+          password: this.state.password,
+        }),
+      };
+      const res = await fetch(`/auth/login`, requestOptions);
+      const msg = await res.text();
+      if (msg === "successfully logged in") {
+        this.props.onAuthChange(true);
+        this.props.history.push("/streams");
+      } else {
+        console.log(msg);
```
Comment (chc5) [inline src/web/client/src/components/Login/Login.tsx:63]: Shouldn't this tell the user something about login failed?

File: src/web/client/src/components/Navbar/Navbar.test.tsx (hunk 1 above)
Comment (chc5) [reply src/web/client/src/components/Navbar/Navbar.test.tsx:18]: ikr

File: src/web/client/src/components/Navbar/Navbar.test.tsx [hunk 4]
```diff
@@ -0,0 +1,55 @@
...
+
+describe("Navbar component", () => {
+  it("should render the Navbar component", () => {
+    const onAuthChangeMock: jest.Mock = jest.fn();
+    const { container } = render(
+      <Router>
+        <Navbar isLoggedIn={false} onAuthChange={onAuthChangeMock} />
+      </Router>,
+    );
+    expect(container.firstChild).toMatchSnapshot();
```
Comment (chc5) [inline src/web/client/src/components/Navbar/Navbar.test.tsx:15]: Use shallow for this

File: src/web/client/src/components/ProtectedRoute/ProtectedRoute.test.tsx [hunk 5]
```diff
@@ -0,0 +1,30 @@
...
+      <ProtectedRoute isLoggedIn={false} redirectPath={"/"} />,
+    );
+    expect(wrapper.find("Redirect")).toHaveLength(1);
+  });
+
+  it("renders Route when isLoggedIn is true", () => {
+    const wrapper: any = shallow(
+      <ProtectedRoute isLoggedIn={true} redirectPath={"/"} />,
+    );
+    expect(wrapper.find("Route")).toHaveLength(1);
```
Comment (chc5) [inline src/web/client/src/components/ProtectedRoute/ProtectedRoute.test.tsx:28]: Why are you comparing length? What is the exact path does it route to?

File: src/web/client/src/components/ProtectedRoute/ProtectedRoute.test.tsx [hunk 6]
```diff
@@ -0,0 +1,30 @@
...
+      </Router>,
+    );
+    expect(container.firstChild).toMatchSnapshot();
+  });
+
+  it("renders Redirect when isLoggedIn is false", () => {
+    const wrapper: any = shallow(
+      <ProtectedRoute isLoggedIn={false} redirectPath={"/"} />,
+    );
+    expect(wrapper.find("Redirect")).toHaveLength(1);
```
Comment (chc5) [inline src/web/client/src/components/ProtectedRoute/ProtectedRoute.test.tsx:21]: Why are you comparing length? What is the exact path does it route to?

File: src/web/client/src/components/Register/Register.test.tsx [hunk 7]
```diff
@@ -0,0 +1,75 @@
...
+import { shallow } from "enzyme";
+import { render } from "@testing-library/react";
+import { BrowserRouter as Router } from "react-router-dom";
+import Register from "./Register";
+
+describe("Register component", () => {
+  let registerWrapper: any;
+  let onAuthChangeMock: jest.Mock;
+  let historyMock: object;
+  let eventMock: any;
```
Comment (chc5) [inline src/web/client/src/components/Register/Register.test.tsx:11]: eventmock is an object

File: src/web/client/src/components/Register/Register.test.tsx [hunk 8]
```diff
@@ -0,0 +1,75 @@
+import React, { FormEvent } from "react";
+import { shallow } from "enzyme";
+import { render } from "@testing-library/react";
+import { BrowserRouter as Router } from "react-router-dom";
+import Register from "./Register";
+
+describe("Register component", () => {
+  let registerWrapper: any;
```
Comment (chc5) [inline src/web/client/src/components/Register/Register.test.tsx:8]: ShallowWrapper

File: src/web/client/src/components/Register/Register.test.tsx [hunk 9]
```diff
@@ -0,0 +1,75 @@
...
+  let onAuthChangeMock: jest.Mock;
+  let historyMock: object;
+  let eventMock: any;
+  let preventDefaultMock: jest.Mock;
+  beforeEach(() => {
+    // Mock Fetch
+    const mockSuccessResponse = "successfully registered";
+    const mockTextPromise: any = Promise.resolve(mockSuccessResponse);
+    const mockFetchPromise: any = Promise.resolve({
```
Comment (chc5) [inline src/web/client/src/components/Register/Register.test.tsx:17]: Promise<string> and Promise

File: src/web/client/src/components/Register/Register.tsx [hunk 10]
```diff
@@ -40,34 +39,34 @@ export default class Register extends Component<Props, State> {
...
+      if (msg === "Bad username") {
+        console.log("no good");
+      }
+      // Successfully registered
+      else {
+        this.props.onAuthChange(true);
+        this.props.history.push("/streams");
+      }
+    } catch (error) {
+      console.log(error);
```
Comment (chc5) [inline src/web/client/src/components/Register/Register.tsx:68]: What happens if there's an error? Warn the user on screen if possible

File: src/web/client/src/components/Navbar/Navbar.test.tsx (hunk 1 above)
Comment (yliu0715) [reply src/web/client/src/components/Navbar/Navbar.test.tsx:18]: yes

File: src/web/client/src/components/ProtectedRoute/ProtectedRoute.test.tsx (hunk 5 above)
Comment (yliu0715) [reply src/web/client/src/components/ProtectedRoute/ProtectedRoute.test.tsx:28]: I'm basically just verifying that the Route component is rendered if isLoggedIn is set to true.
https://stackoverflow.com/a/44426403

File: src/web/client/src/components/ProtectedRoute/ProtectedRoute.test.tsx (hunk 6 above)
Comment (yliu0715) [reply src/web/client/src/components/ProtectedRoute/ProtectedRoute.test.tsx:21]: See above

File: src/web/client/src/components/LiveStream/LiveStream.test.tsx (hunk 2 above)
Comment (yliu0715) [reply src/web/client/src/components/LiveStream/LiveStream.test.tsx:9]: I thought so too, but I couldn't find a right typing for this.

File: src/web/client/src/components/Register/Register.test.tsx (hunk 8 above)
Comment (yliu0715) [reply src/web/client/src/components/Register/Register.test.tsx:8]: Typescript couldn't find a ShallowWrapper type

File: src/web/client/src/components/Register/Register.test.tsx (hunk 7 above)
Comment (yliu0715) [reply src/web/client/src/components/Register/Register.test.tsx:11]: It gives me `Property 'target' does not exist on type 'object'.`

File: src/web/client/src/components/LiveStream/LiveStream.test.tsx (hunk 2 above)
Comment (yliu0715) [reply src/web/client/src/components/LiveStream/LiveStream.test.tsx:9]: It gives me this `Generic type 'Promise<T>' requires 1 type argument(s).`

File: src/web/client/src/components/Register/Register.test.tsx (hunk 9 above)
Comment (yliu0715) [reply src/web/client/src/components/Register/Register.test.tsx:17]: It gives me this `Generic type 'Promise<T>' requires 1 type argument(s).`

File: src/web/client/src/components/LiveStream/LiveStream.test.tsx (hunk 2 above)
Comment (melvinheks) [reply src/web/client/src/components/LiveStream/LiveStream.test.tsx:9]: what about Promise<any> :)

Comment (yliu0715) [reply src/web/client/src/components/LiveStream/LiveStream.test.tsx:9]: yes

**Pull request merged**